src/adapter/src/catalog/config.rs

```diff
@@ -56,6 +57,10 @@ pub struct Config<'a, S> {
     pub secrets_reader: Arc<dyn SecretsReader>,
     /// IP Addresses which will be used for egress.
     pub egress_ips: Vec<Ipv4Addr>,
+    /// A optional frontend used to pull system parameters for initial sync in
+    /// Catalog::open. A `None` value indicates that the initial sync should be
+    /// skipped.
+    pub system_parameter_frontend: Option<Arc<SystemParameterFrontend>>,
```

Comment:
This seems like exactly the kind of coupling between components we were hoping to avoid? Gut take is that we should either avoid this coupling, or accept the coupling and use that to simplify the code. But having half of the code coupled and half of the code decoupled seems like we wind up with the downsides of both.

IMO, seems easiest to accept the coupling and move the code into the adapter directly. Wouldn't need to deal with `tokio_postgres` and the network failures, and you'd have access to the `ServerVar` defaults.

Reply:
> This seems like exactly the kind of coupling between components we were hoping to avoid?

Unless we introduce some form hook that can be used to extend the catalog state initialization there is no way to avoid this coupling. We need the `config_has_synced_once` value to determine whether to override the state or not, so something like:

```rust
trait InitCatalogState {
    type E: Into<AdapterError>
    async fn state(state: &mut CatalogState) -> Result<(), E>;
}
```
and then pass an `InitCatalogState` type instead of the `system_parameter_frontend` in all `Config`.

However, it seems more pragmatic to just move everything under `adapter` for now...

Reply:
> IMO, seems easiest to accept the coupling and move the code into the adapter directly

I did that in the last two commits of the updated PR.